foyer-memory/src/cache.rs

```diff
@@ -226,64 +220,177 @@ where
-where
-    K: Key,
-    V: Value,
-    L: CacheEventListener<K, V>,
-    S: BuildHasher + Send + Sync + 'static,
-{
-    fn from(value: FifoCacheConfig<K, V, L, S>) -> Self {
-        Self::Fifo(value)
+impl From<FifoConfig> for EvictionConfig {
+    fn from(value: FifoConfig) -> EvictionConfig {
+        EvictionConfig::Fifo(value)
     }
 }
 
-impl<K, V, L, S> From<LruCacheConfig<K, V, L, S>> for CacheConfig<K, V, L, S>
+impl From<LruConfig> for EvictionConfig {
+    fn from(value: LruConfig) -> EvictionConfig {
+        EvictionConfig::Lru(value)
+    }
+}
+
+impl From<LfuConfig> for EvictionConfig {
+    fn from(value: LfuConfig) -> EvictionConfig {
+        EvictionConfig::Lfu(value)
+    }
+}
+
+impl From<S3FifoConfig> for EvictionConfig {
+    fn from(value: S3FifoConfig) -> EvictionConfig {
+        EvictionConfig::S3Fifo(value)
+    }
+}
+
+pub struct CacheBuilder<K, V, L, S>
 where
     K: Key,
     V: Value,
     L: CacheEventListener<K, V>,
     S: BuildHasher + Send + Sync + 'static,
 {
-    fn from(value: LruCacheConfig<K, V, L, S>) -> Self {
-        Self::Lru(value)
-    }
+    capacity: usize,
+    shards: Option<usize>,
+    eviction_config: Option<EvictionConfig>,
+    object_pool_capacity: Option<usize>,
+    event_listener: L,
+    hash_builder: S,
+    _marker: PhantomData<(K, V)>,
 }
 
-impl<K, V, L, S> From<LfuCacheConfig<K, V, L, S>> for CacheConfig<K, V, L, S>
+impl<K, V> CacheBuilder<K, V, DefaultCacheEventListener<K, V>, RandomState>
 where
     K: Key,
     V: Value,
-    L: CacheEventListener<K, V>,
-    S: BuildHasher + Send + Sync + 'static,
 {
-    fn from(value: LfuCacheConfig<K, V, L, S>) -> Self {
-        Self::Lfu(value)
+    pub fn new(capacity: usize) -> Self {
+        Self {
+            capacity,
+            shards: None,
+            eviction_config: None,
+            object_pool_capacity: None,
+            event_listener: DefaultCacheEventListener::default(),
+            hash_builder: RandomState::default(),
+            _marker: PhantomData,
+        }
     }
 }
 
-impl<K, V, L, S> From<S3FifoCacheConfig<K, V, L, S>> for CacheConfig<K, V, L, S>
+impl<K, V, L, S> CacheBuilder<K, V, L, S>
 where
     K: Key,
     V: Value,
     L: CacheEventListener<K, V>,
     S: BuildHasher + Send + Sync + 'static,
 {
-    fn from(value: S3FifoCacheConfig<K, V, L, S>) -> Self {
-        Self::S3Fifo(value)
+    const DEFAULT_SHARDS: usize = 1;
+    const DEFAULT_EVICTION_CONFIG: EvictionConfig = EvictionConfig::Lfu(LfuConfig {
+        window_capacity_ratio: 0.1,
+        protected_capacity_ratio: 0.8,
+        cmsketch_eps: 0.001,
+        cmsketch_confidence: 0.9,
+    });
+    const DEFAULT_OBJECT_POOL_CAPACITY_RATIO_RECIPROCAL: usize = 10;
+
+    pub fn with_shards(mut self, shards: usize) -> Self {
+        self.shards = Some(shards);
+        self
+    }
+
+    pub fn with_eviction_config(mut self, eviction_config: EvictionConfig) -> Self {
```

Comment:
Can even be `fn with_eviction_config(mut self: eviction_config: impl Into<EvictionConfig>)`.

Reply:
Sensei!!!